Widgets/CodeWidget.h

```diff
@@ -1,16 +1,95 @@
 #ifndef CODEWIDGET_H
 #define CODEWIDGET_H
 
+#include <QFileDialog>
 #include <QFont>
+#include <QInputDialog>
+#include <QMessageBox>
+#include <QPrinter>
+#include <QTextStream>
 #include <QWidget>
 
 class CodeWidget : public QWidget {
+  Q_OBJECT
+  Q_PROPERTY(QString code READ code WRITE setCode USER true)
+
  public:
   explicit CodeWidget(QWidget* parent);
   ~CodeWidget();
 
+  QString code() const;
+  void setCode(QString);
+  int lineCount();
+
+ public slots:
+  void newSource() {
+    QMessageBox::StandardButton reply;
+    reply =
+        QMessageBox::question(this, tr("New Source"), tr("Are you sure you want to clear the source and start over?"),
+                              QMessageBox::Yes | QMessageBox::No);
+    if (reply == QMessageBox::Yes) {
+      this->setCode("");
+    }
+  }
+
+  void loadSource() {
+    QFileDialog fileDialog(this);
+    fileDialog.setWindowModality(Qt::WindowModal);
+    fileDialog.setAcceptMode(QFileDialog::AcceptOpen);
+    fileDialog.setFileMode(QFileDialog::ExistingFile);
+    fileDialog.setMimeTypeFilters(this->fileFilters());
+    if (fileDialog.exec() != QDialog::Accepted) return;
+    const QString fn = fileDialog.selectedFiles().first();
+    if (fn.isEmpty()) return;
+    QFile file(fn);
+
+    file.open(QIODevice::ReadOnly | QIODevice::Text);
+    QTextStream in(&file);
+    this->setCode(in.readAll());
+  }
+
+  void saveSource() {
+    QFileDialog fileDialog(this);
+    fileDialog.setWindowModality(Qt::WindowModal);
+    fileDialog.setAcceptMode(QFileDialog::AcceptSave);
+    fileDialog.setMimeTypeFilters(this->fileFilters());
+    if (fileDialog.exec() != QDialog::Accepted) return;
+
+    const QString fn = fileDialog.selectedFiles().first();
+    if (fn.isEmpty()) return;
+    QFile file(fn);
+
+    file.open(QIODevice::WriteOnly | QIODevice::Text);
+    QTextStream out(&file);
+    out << this->code();
+  }
+
+  void printSource();
+
+  void gotoLineDialog() {
+    bool ok;
+    int lineNumber = QInputDialog::getInt(this, tr("Go to Line"), tr("Line:"), 1, 1, lineCount(), 1, &ok);
+    if (!ok) return;
+    gotoLine(lineNumber);
+  }
+
+  void print(QPrinter* printer);
+  void undo();

```

Comment:
As we discussed over IRC, CRTP does not really fit with the current design because it's a "uses a" instead of a "is a" relationship. Further, it doesn't matter if it did because this is a particular area we want to leave open to IDE plugins, so one could hypothetically do something like embed VS Code's Monaco Editor in RGM in place of QScintilla. CRTP, as Josh told me, is difficult to get working across binaries.